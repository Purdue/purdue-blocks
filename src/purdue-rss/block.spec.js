// title-nav test
jest.setTimeout(30000)
import {
    createNewPost,
    enablePageDialogAccept,
    getEditedPostContent,
    insertBlock,
} from '@wordpress/e2e-test-utils';

import {
    selectOption,
    clickElementByText,
    clickElementByAria,
    blockStartup,
    updateRangeInput,
    clickCheckbox,
    clickRadio,
    openSidebarPanelWithTitle
} from '../test-helpers'

const block = {blockTitle: 'Purdue RSS feed', blockName: 'purdue-blocks/purdue-rss'}

describe( 'ðŸ”¬ Purdue RSS Block', () => {
    beforeAll( async () => {
        await enablePageDialogAccept();
    } );
    beforeEach( async () => {
        await createNewPost();
    } );

    test( 'ðŸ”Ž Block should be available.', async () => {
        await insertBlock( 'Purdue RSS feed' )


        // tests that the block is properly inserted and matches the existing snapshot
        expect(await page.$('[data-type="purdue-blocks/purdue-rss"]')).not.toBeNull()
        expect( await getEditedPostContent() ).toMatchSnapshot()
    } )

    describe('ðŸ”¬ Block Editor Fields', () => {
        describe('ðŸ”¬ Feed URL Editor', () => {
            var originalTimeout;
            test('ðŸ”Ž Feed URL is editable', async () => {
                await blockStartup(block)

                const typeString = "https://www.purdue.edu/"
    
                await page.focus('input[placeholder="Enter URL hereâ€¦"]')
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"feedURL":"${typeString}"`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”Ž Feed URL can be submitted', async () => {
                await blockStartup(block)
                const url = "https://www.purdue.edu/"
    
                await page.focus('input[placeholder="Enter URL hereâ€¦"]')
                await page.keyboard.type(url, {delay: 10})
                await clickElementByText('button', 'Use URL')
                const typeString = "Header text"   
                await page.focus( '.block-editor-rich-text__editable.feed-header' );
                await page.keyboard.type(typeString, {delay: 10})

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"error":"Invalid RSS feed URL."`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
        })
        describe('ðŸ”¬ Feeds Display Editor', () => {
            test('ðŸ”Ž Header field should be editable', async () => {
                await blockStartup(block)
                const url = "https://www.purdue.edu/"
    
                await page.focus('input[placeholder="Enter URL hereâ€¦"]')
                await page.keyboard.type(url, {delay: 10})
                await clickElementByText('button', 'Use URL')

                const typeString = "Header text"   
                await page.focus( '.block-editor-rich-text__editable.feed-header' );
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()

                expect( editedContent.includes(`<p class="feed-header">${typeString}</p>`)).toBe(true)
                expect( editedContent).toMatchSnapshot()
            })
            test('ðŸ”Ž Pencil on the toolbar is clickable', async () => {
                await blockStartup(block)

                const typeString = "https://www.purdue.edu/"
    
                await page.focus('input[placeholder="Enter URL hereâ€¦"]')
                await page.keyboard.type(typeString, {delay: 10})
                await clickElementByText('button', 'Use URL')
                await page.focus('div[aria-label="Block: Purdue RSS feed"]')
                await clickElementByAria('button', 'Edit RSS URL')
    
                expect(await page.$(`div.news-feed-editor`)).not.toBeNull()
            })
            test( 'ðŸ”Ž Button should open Media Library.', async () => {
                await blockStartup(block)
                const url = "http://lorem-rss.herokuapp.com/feed"
    
                await page.focus('input[placeholder="Enter URL hereâ€¦"]')
                await page.keyboard.type(url, {delay: 10})
                await clickElementByText('button', 'Use URL')

                const typeString = "Header text"   
                await page.focus( '.block-editor-rich-text__editable.feed-header' );
                await page.keyboard.type(typeString, {delay: 10})
                await clickRadio('Layout of the RSS feed', 'withoutImage')
                // open media library
                await clickElementByText('button', 'Select an image')
    
                // tests that media library does open
                expect(await page.$(`div.media-frame-title`)).not.toBeNull()
            })
        })
        describe( 'ðŸ”¬ Side Panel Settings', () => {
            describe( 'ðŸ”¬ Layout of the RSS feed should be selectable', () => {
                test( 'ðŸ”Ž Recent news without image should be selectable.', async () => {
                    await blockStartup(block)
        
                    // select the radio button with the virtual keyboard
                    await clickRadio('Layout of the RSS feed', 'withoutImage')
        
                    const editedContent = await getEditedPostContent()
        
                    expect( editedContent.includes(`"type":"withoutImage"`)).toBe(true)
                    expect( editedContent).toMatchSnapshot()
                })
                test( 'ðŸ”Ž All News/Events should be selectable.', async () => {
                    await blockStartup(block)
        
                    // select the radio button with the virtual keyboard
                    await clickRadio('Layout of the RSS feed', 'all')
        
                    const editedContent = await getEditedPostContent()
        
                    expect( editedContent.includes(`"type":"all"`)).toBe(true)
                    expect( editedContent).toMatchSnapshot()
                })
            })
            test( 'Heading level dropdown selector', async () => {
                await blockStartup(block)
    
                await selectOption('Heading level of the Header', 'h2')
                let editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h2"`)).toBe(true)
    
                await selectOption('Heading level of the Header', 'h3')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h3"`)).toBe(true)
    
                await selectOption('Heading level of the Header', 'h4')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h4"`)).toBe(true)
    
                await selectOption('Heading level of the Header', 'h5')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h5"`)).toBe(true)
    
                await selectOption('Heading level of the Header', 'h6')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h6"`)).toBe(true)
    
                await selectOption('Heading level of the Header', 'p')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h2"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h3"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h4"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h5"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h6"`)).toBe(false)
    
            })
            test( 'ðŸ”Ž Image alt text should be editable.', async () => {
                await blockStartup(block)
    
                const typeString = "Image alt text."
                
                // focus the text box then type into it with the virtual keyboard
                await clickElementByText('label', 'Image Alt Text')
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()
    
                // tests: 
                // a. save output properly contains the correct string 
                // b. save output matches the existing snapshot
                expect( editedContent.includes(`"altText":"${typeString}"`)).toBe(true)
                // expect( editedContent).toMatchSnapshot()
            })
            describe( 'ðŸ”¬ Link can be added', () => {
                test( 'ðŸ”Ž "Add a link to the news/events page?" checkbox reveals link inputs', async () => {
                    await blockStartup(block)
    
                    await clickCheckbox('Add a link to the news/events page?')
                          
                    let editedContent = await getEditedPostContent()
                    
                    expect( editedContent.includes('{"hasLink":true}')).toBe(true)
                    expect( await getEditedPostContent() ).toMatchSnapshot()
     
                    await clickCheckbox('Add a link to the news/events page?')
                    
                    editedContent = await getEditedPostContent()
                    
                    expect( editedContent.includes('{"hasLink":true}')).toBe(false)
                    expect( await getEditedPostContent() ).toMatchSnapshot()
                })
                test('ðŸ”Ž Link text field updates output correctly', async () => {
                    await blockStartup(block)
    
                    const typeString = "CTA text"
                    await clickCheckbox('Add a link to the news/events page?')
                    await clickElementByText('label', 'Link text')
                    await page.keyboard.type(typeString, {delay: 10})
    
                    const editedContent = await getEditedPostContent()
                    expect( editedContent.includes(`"linkText":"${typeString}"`)).toBe(true)
                    expect( await getEditedPostContent() ).toMatchSnapshot()
                })
                test('ðŸ”Ž Link url field updates output correctly', async () => {
                    await blockStartup(block)
                    await clickCheckbox('Add a link to the news/events page?')
                    const typeString = "https://www.purdue.edu"
    
                    await clickElementByText('label', 'Link URL')
                    await page.keyboard.type(typeString, {delay: 10})
    
                    const editedContent = await getEditedPostContent()
                    expect( editedContent.includes(`"link":"${typeString}"`)).toBe(true)
                    expect( await getEditedPostContent() ).toMatchSnapshot()
                })
                test('ðŸ”Ž Link can be opened in new tab', async () => {
                    await blockStartup(block)
    
                    const typeString = "https://www.purdue.edu"
                    await clickCheckbox('Add a link to the news/events page?')
                    await clickCheckbox('Open link in new tab?')
    
                    const editedContent = await getEditedPostContent()
                    expect( editedContent.includes(`"external": false`)).toBe(false)
                    expect( await getEditedPostContent() ).toMatchSnapshot()
                })
            })
        })
    })
})