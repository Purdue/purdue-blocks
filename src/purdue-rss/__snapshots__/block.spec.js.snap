// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🔬 Purdue RSS Block 🔎 Block should be available. 1`] = `
"<!-- wp:purdue-blocks/purdue-rss -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Feed URL Editor 🔎 Feed URL can be submitted 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"feedURL\\":\\"https://www.purdue.edu/\\",\\"data\\":null,\\"error\\":\\"Invalid RSS feed URL.\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"feed-header\\">Header text</p><p class=\\"error\\">Invalid RSS feed URL.</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Feed URL Editor 🔎 Feed URL is editable 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"feedURL\\":\\"https://www.purdue.edu/\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Feeds Display Editor 🔎 Header field should be editable 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"feedURL\\":\\"https://www.purdue.edu/\\",\\"data\\":null,\\"error\\":\\"Invalid RSS feed URL.\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"feed-header\\">Header text</p><p class=\\"error\\">Invalid RSS feed URL.</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Layout of the RSS feed should be selectable 🔎 All News/Events should be selectable. 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"type\\":\\"all\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Layout of the RSS feed should be selectable 🔎 Recent news without image should be selectable. 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"type\\":\\"withoutImage\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Link can be added 🔎 "Add a link to the news/events page?" checkbox reveals link inputs 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"hasLink\\":true} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Link can be added 🔎 "Add a link to the news/events page?" checkbox reveals link inputs 2`] = `
"<!-- wp:purdue-blocks/purdue-rss -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Link can be added 🔎 Link can be opened in new tab 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"hasLink\\":true,\\"external\\":true} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Link can be added 🔎 Link text field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"hasLink\\":true,\\"linkText\\":\\"CTA text\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;

exports[`🔬 Purdue RSS Block 🔬 Block Editor Fields 🔬 Side Panel Settings 🔬 Link can be added 🔎 Link url field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/purdue-rss {\\"hasLink\\":true,\\"link\\":\\"https://www.purdue.edu\\"} -->
<div class=\\"news-feed\\"><div class=\\"container\\"><p class=\\"error\\">Loading...</p></div></div>
<!-- /wp:purdue-blocks/purdue-rss -->"
`;
